static/js/Misc.jsx

```diff
@@ -2056,110 +2056,331 @@ SignUpModal.propTypes = {
 };
 
+// Have a callback function run when its children (element or nested group of elements) become fully visible within the viewport
+function OnInView({ children, onVisible }) {
+  // Get a mutable reference object for the child and/or children to be rendered within this component wrapped in a div
+  // The reference is attached to the DOM element returned from this component
+  // This allows us to access properties of the DOM element directly
+  const elementRef = useRef(); 
 
-class InterruptingMessage extends Component {
-  constructor(props) {
-    super(props);
-    this.displayName = 'InterruptingMessage';
-    this.state = {
-      timesUp: false,
-      animationStarted: false
-    };
-    this.settings = {
-      "modal": {
-        "trackingName": "Interrupting Message",
-        "showDelay": 1000,
+  useEffect(() => {
+    const observer = new IntersectionObserver(
+      // Callback function will be invoked whenever the visibility of the observed element changes
+      (entries) => {
+        const [entry] = entries;
+        // Check if the observed element is intersecting with the viewport (it's visible)
+        // Invoke provided prop callback for analytics purposes
+        if (entry.isIntersecting) {
+          onVisible();
+        }
       },
-      "banner": {
-        "trackingName": "Banner Message",
-        "showDelay": 1,
-      }
-    }[this.props.style];
-  }
-  componentDidMount() {
-    if (this.shouldShow()) {
-      this.delayedShow();
+      // The entire element must be entirely visible
+      { threshold: 1 }
+    );
+
+    // Start observing the element, but wait until the element exists
+    if (elementRef.current) {
+      observer.observe(elementRef.current);
     }
-  }
-  shouldShow() {
+
+    // Cleanup when the component unmounts
+    return () => {
+      // Stop observing the element when it's no longer on the screen and can't be visible
+      if (elementRef.current) {
+        observer.unobserve(elementRef.current);
+      }
+    };
+  }, [onVisible]);
+
+  // Attach elementRef to a div wrapper and pass the children to be rendered within it
+  return <div ref={elementRef}>{children}</div>;
+}
+
+
+const InterruptingMessage = ({
+  onClose,
+}) => {
+  const [interruptingMessageShowDelayHasElapsed, setInterruptingMessageShowDelayHasElapsed] = useState(false);
+  const [hasInteractedWithModal, setHasInteractedWithModal] = useState(false);
+  const strapi = useContext(StrapiDataContext);
+
+  const markModalAsHasBeenInteractedWith = (modalName) => {
+    localStorage.setItem("modal_" + modalName, "true");
+  };
+
+  const hasModalBeenInteractedWith = (modalName) => {
+    return JSON.parse(localStorage.getItem("modal_" + modalName));
+  };
+
+  const trackModalInteraction = (modalName, eventDescription) => {
+    gtag("event", "modal_interacted_with_" + eventDescription, {
+      campaignID: modalName,
+      adType: "modal",
+    });
+  };
+
+  const trackModalImpression = () => {
```

Comment:
I'm wondering if the analytics functions should not be moved out to a dedicated analytics module and be triggered from there. In a sense they are not part of the core functionality of the components.

Reply:
I think they should but a larger discussion with everyone should be had to decide how that should be done, but that isn't my domain. I personally think all calls to the gtag function should be put under an adapter class/module. That way if Sefaria changes its analytics platform to something else besides GA there won't have to be as many changes to the code